### PR TITLE
chore: move to ESM, update all dependencies

### PR DESCRIPTION
**Pull request by Trott**
`chalk` requires ESM. Sure, let's do it.

`nopt` doesn't support Node.js 12.x so this is a breaking change to drop 12.x support. (I guess the ESM stuff might also be a breaking change for people using this as a dependency rather than a CLI. Does anyone do that?)

### CODE REVIEW

**Comment by Trott** [comment]
@nodejs/tsc Anyone willing to review this?

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [12.x, 14.x, 16.x, 17.x, 18.x, 19.x]
+        node-version: [14.x, 16.x, 17.x, 18.x, 19.x]
```
Comment (targos) [inline .github/workflows/node.js.yml:19]:
```suggestion
        node-version: [14.x, 16.x, 18.x, 19.x]
```

File: bin/cmd.mjs [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
-const Validator = require('../lib')
-const Tap = require('../lib/tap')
-const utils = require('../lib/utils')
-const subsystem = require('../lib/rules/subsystem')
+import { exec } from 'child_process'
+import fs from 'fs'
+import http from 'http'
+import https from 'https'
+import nopt from 'nopt'
+import path from 'path'
```
Comment (targos) [inline bin/cmd.mjs:8]:
```suggestion
import { exec } from 'node:child_process'
import fs from 'node:fs'
import http from 'node:http'
import https from 'node:https'
import path from 'node:path'
import nopt from 'nopt'
```

File: lib/tap.mjs [hunk 3]
```diff
@@ -1,6 +1,5 @@
-'use strict'
-
-const util = require('util')
+import util from 'util'
+import { Readable } from 'stream'
```
Comment (targos) [inline lib/tap.mjs:2]:
```suggestion
import util from 'node:util'
import { Readable } from 'node:stream'
```

File: lib/utils.mjs [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import chalk from 'chalk'
+
+export const CHECK = chalk.green('✔')
+export const X = chalk.red('✖')
+export const WARN = chalk.yellow('⚠')
+
+export const rightPad = function rightPad (str, max) {
```
Comment (targos) [inline lib/utils.mjs:7]:
```suggestion
export function rightPad (str, max) {
```

File: lib/utils.mjs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  const diff = max - str.length + 1
+  if (diff > 0) {
+    return `${str}${' '.repeat(diff)}`
+  }
+  return str
+}
+
+export const leftPad = function leftPad (str, max) {
```
Comment (targos) [inline lib/utils.mjs:15]:
```suggestion
export function leftPad (str, max) {
```

File: lib/utils.mjs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  const diff = max - str.length + 1
+  if (diff > 0) {
+    return `${' '.repeat(diff)}${str}`
+  }
+  return str
+}
+
+export const header = (sha, status) => {
```
Comment (targos) [inline lib/utils.mjs:23]:
```suggestion
export function header (sha, status) {
```

File: lib/utils.mjs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    case 'pass': {
+      const suffix = status === 'skip' ? ' # SKIPPED' : ''
+      return `${CHECK}  ${chalk.underline(sha)}${suffix}`
+    }
+    case 'fail':
+      return `${X}  ${chalk.underline(sha)}`
+  }
+}
+
+export const describeRule = function describeRule (rule, max = 20) {
```
Comment (targos) [inline lib/utils.mjs:35]:
```suggestion
export function describeRule (rule, max = 20) {
```

File: lib/utils.mjs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  if (rule.meta && rule.meta.description) {
+    const desc = rule.meta.description
+    const title = leftPad(rule.id, max)
+    console.log(' %s %s', chalk.red(title), chalk.dim(desc))
+  }
+}
+
+export const describeSubsystem = function describeSubsystem (subsystems, max = 20) {
```
Comment (targos) [inline lib/utils.mjs:43]:
```suggestion
export function describeSubsystem (subsystems, max = 20) {
```

File: bin/cmd.mjs [hunk 9]
```diff
@@ -1,19 +1,18 @@
...
+import pretty from '../lib/format-pretty.mjs'
+import formatTap from '../lib/format-tap.mjs'
+import Validator from '../lib/validator.mjs'
+import Tap from '../lib/tap.mjs'
+import utils from '../lib/utils.mjs'
```
Comment (targos) [inline bin/cmd.mjs:13]:
```suggestion
import * as utils from '../lib/utils.mjs'
```

File: lib/format-pretty.mjs [hunk 10]
```diff
@@ -1,11 +1,9 @@
-'use strict'
-
-const chalk = require('chalk')
-const utils = require('./utils')
+import chalk from 'chalk'
+import utils from './utils.mjs'
```
Comment (targos) [inline lib/format-pretty.mjs:2]:
```suggestion
import * as utils from './utils.mjs'
```

File: lib/utils.mjs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+export default {
+  CHECK,
+  X,
+  WARN,
+  rightPad,
+  leftPad,
+  header,
+  describeRule,
+  describeSubsystem
+}
```
Comment (targos) [inline lib/utils.mjs:64]:
```suggestion
```

File: test/cli-test.mjs [hunk 12]
```diff
@@ -1,13 +1,13 @@
...
+import { test } from 'tap'
+import { readFileSync } from 'fs'
+import { spawn } from 'child_process'
+import subsystems from '../lib/rules/subsystem.mjs'
 
-const { test } = require('tap')
-const { readFileSync } = require('fs')
-const { spawn } = require('child_process')
-const subsystems = require('../lib/rules/subsystem')
+import path from 'path'
```
Comment (targos) [inline test/cli-test.mjs:6]:
```suggestion
import path from 'node:path'
```

File: test/cli-test.mjs [hunk 13]
```diff
@@ -1,13 +1,13 @@
-'use strict'
+import { test } from 'tap'
+import { readFileSync } from 'fs'
+import { spawn } from 'child_process'
```
Comment (targos) [inline test/cli-test.mjs:3]:
```suggestion
import { readFileSync } from 'node:fs'
import { spawn } from 'node:child_process'
```

File: test/cli-test.mjs [hunk 14]
```diff
@@ -143,9 +143,12 @@ test('Test cli flags', (t) => {
       tt.fail('This should not happen')
     })
 
-    ls.on('close', (code) => {
+    ls.on('close', async (code) => {
+      const pkgJsonPath = path.join(new URL(import.meta.url).pathname, '../../package.json')
```
Comment (targos) [inline test/cli-test.mjs:147]:
```suggestion
      const pkgJsonPath = new URL('../package.json', import.meta.url)
```

`url.pathname` doesn't work fine on Windows. URL objects can be passed directly to `fs` methods.

File: test/utils-test.mjs [hunk 15]
```diff
@@ -1,7 +1,5 @@
-'use strict'
-
-const { test } = require('tap')
-const utils = require('../lib/utils')
+import { test } from 'tap'
+import utils from '../lib/utils.mjs'
```
Comment (targos) [inline test/utils-test.mjs:2]:
```suggestion
import * as utils from '../lib/utils.mjs'
```

File: test/validator.mjs [hunk 16]
```diff
@@ -1,7 +1,7 @@
-'use strict'
-
-const test = require('tap').test
-const Validator = require('../')
+import { test } from 'tap'
+import Validator from '../lib/validator.mjs'
+import { readFileSync } from 'fs'
+import { join } from 'path'
```
Comment (targos) [inline test/validator.mjs:4]:
```suggestion
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
```

File: test/validator.mjs [hunk 17]
```diff
@@ -188,14 +188,16 @@ test('Validator - misc', (t) => {
 
 test('Validator - real commits', (t) => {
   t.test('basic', (tt) => {
+    const commit = JSON.parse(readFileSync(join(new URL(import.meta.url).pathname, '..', 'fixtures', 'commit.json'), { encoding: 'utf8' }))
+    const pr = JSON.parse(readFileSync(join(new URL(import.meta.url).pathname, '..', 'fixtures', 'pr.json'), { encoding: 'utf8' }))
```
Comment (targos) [inline test/validator.mjs:192]:
```suggestion
    const commit = JSON.parse(readFileSync(new URL('fixtures/commit.json', import.meta.url), { encoding: 'utf8' }))
    const pr = JSON.parse(readFileSync(new URL('fixtures/pr.json', import.meta.url), { encoding: 'utf8' }))
```

File: lib/validator.mjs [hunk 18]
```diff
@@ -1,11 +1,33 @@
-'use strict'
+import EE from 'events'
```
Comment (targos) [inline lib/validator.mjs:1]:
```suggestion
import EE from 'node:events'
```

**Review by targos — APPROVED** [review]

**Comment by Trott** [comment]
The coverage report in tests is now broken. Maybe `tap` doesn't support ESM coverage? (I haven't looked. I'm just guessing.) Replacing with `c8` should work if all else fails. Can do that in a subsequent PR. Or I can do it in this PR if people are confident it's the right way to go. Coverage has been advisory-only anyway. 

Old output:

```
> tap --no-check-coverage --coverage-report=text-summary


=============================== Coverage summary ===============================
Statements   : 92.79% ( 451/486 )
Branches     : 76.01% ( 187/246 )
Functions    : 94.2% ( 65/69 )
Lines        : 93.93% ( 434/462 )
================================================================================
```

Output with `c8`:

```
-------------------|---------|----------|---------|---------|-------------------
File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-------------------|---------|----------|---------|---------|-------------------
All files          |    94.7 |    87.71 |   94.11 |    94.7 |                   
 ...alidate-commit |     100 |      100 |     100 |     100 |                   
  index.mjs        |     100 |      100 |     100 |     100 |                   
 ...ate-commit/bin |   91.92 |    81.81 |     100 |   91.92 |                   
  cmd.mjs          |   91.92 |    81.81 |     100 |   91.92 | 86-87,100-110     
 ...ate-commit/lib |   88.91 |       80 |   91.66 |   88.91 |                   
  ...at-pretty.mjs |    87.2 |    52.38 |     100 |    87.2 | ...33-34,43-44,63 
  format-tap.mjs   |    58.1 |    61.53 |      60 |    58.1 | ...30,34-48,63-74 
  rule.mjs         |     100 |      100 |     100 |     100 |                   
  tap.mjs          |   94.06 |    94.11 |   93.33 |   94.06 | 65-66,70-74       
  utils.mjs        |     100 |    90.47 |     100 |     100 | 47-48             
  validator.mjs    |   98.01 |    90.47 |     100 |   98.01 | 90-91             
 ...mmit/lib/rules |   99.26 |    97.87 |     100 |   99.26 |                   
  ...s-trailer.mjs |     100 |      100 |     100 |     100 |                   
  fixes-url.mjs    |   94.79 |    93.33 |     100 |   94.79 | 91-95             
  ...ter-title.mjs |     100 |       80 |     100 |     100 | 17                
  line-length.mjs  |     100 |      100 |     100 |     100 |                   
  metadata-end.mjs |     100 |      100 |     100 |     100 |                   
  pr-url.mjs       |     100 |      100 |     100 |     100 |                   
  reviewers.mjs    |     100 |      100 |     100 |     100 |                   
  subsystem.mjs    |     100 |      100 |     100 |     100 |                   
  title-format.mjs |     100 |      100 |     100 |     100 |                   
  title-length.mjs |     100 |      100 |     100 |     100 |                   
-------------------|---------|----------|---------|---------|-------------------
```

**Comment by tniessen** [comment]
@Trott is there a reason to use `.mjs` instead of setting the module type to ESM in `package.json`?

**Comment by Trott** [comment]
> @Trott is there a reason to use `.mjs` instead of setting the module type to ESM in `package.json`?

Personal preference. I prefer `.mjs` because it is very hard to miss, as opposed to setting things in `package.json` which seems more magical and hiding things to me.

But also: I know I'm in the minority on this, so I'll update it to set things in `package.json`.

**Comment by aduh95** [comment]
> breaking change for people using this as a dependency rather than a CLI. Does anyone do that?

`node-core-utils` does that (and that's the package listed in https://www.npmjs.com/package/core-validate-commit?activeTab=dependents FWIW).

**Review by aduh95 — APPROVED** [review]
Should we drop official support for v14.x at the same time? v14.x EOL is quite close, and that would allow us to switch to `node:test` for the tests in a future PR.

I'm with you for `.mjs`, but no strong feelings.

**Comment by richardlau** [comment]
> > breaking change for people using this as a dependency rather than a CLI. Does anyone do that?
> 
> `node-core-utils` does that (and that's the package listed in https://www.npmjs.com/package/core-validate-commit?activeTab=dependents FWIW).

I thought it did too, but `node-core-utils` executes `core-validate-commit` as a CLI tool: https://github.com/nodejs/node-core-utils/blob/cc4b11c3de20b0c4b28e9973794b51313a7ed766/lib/landing_session.js#L427-L434

**Comment by GeoffreyBooth** [comment]
> But also: I know I’m in the minority on this, so I’ll update it to set things in `package.json`.

I think in this case specifically, it’s better to keep the `.js` extensions on all the files for more legible git history. Yes it seems like git is able to track all the changes across the renames, but why add that complexity? Simply avoiding an unnecessary rename seems safer.

**Comment by Trott** [comment]
> I think in this case specifically, it’s better to keep the `.js` extensions on all the files for more legible git history. Yes it seems like git is able to track all the changes across the renames, but why add that complexity? Simply avoiding an unnecessary rename seems safer.

Prioritizing git history over code clarity seems like the wrong trade-off to me, but views on this sort of thing tend to depend on assumptions, speculation, and personal preference. So I've already conceded and named the files back to `.js`.

File: bin/cmd.js [hunk 19]
```diff
@@ -1,19 +1,18 @@
...
-const nopt = require('nopt')
-const path = require('path')
-const pretty = require('../lib/format-pretty')
-const formatTap = require('../lib/format-tap')
-const Validator = require('../lib')
-const Tap = require('../lib/tap')
-const utils = require('../lib/utils')
-const subsystem = require('../lib/rules/subsystem')
+import { exec } from 'node:child_process'
+import fs from 'node:fs'
```
Comment (GeoffreyBooth) [inline bin/cmd.js:4]: Nit: All of these could be `import fs from 'node:fs/promises'`

File: package.json [hunk 20]
```diff
@@ -3,17 +3,17 @@
   "version": "3.20.0",
   "description": "Validate the commit message for a particular commit in node core",
   "main": "index.js",
```
Comment (GeoffreyBooth) [inline package.json:5]:
```suggestion
  "exports": "index.js",
```

File: package.json [hunk 21]
```diff
@@ -3,17 +3,17 @@
   "version": "3.20.0",
```
Comment (GeoffreyBooth) [inline package.json:3]: This should probably be a major version bump?

File: package.json (hunk 20 above)
Comment (aduh95) [reply package.json:5]: We probably want to also export the `package.json` file.

File: package.json (hunk 21 above)
Comment (Trott) [reply package.json:3]: We do the version bumps when we release.

**Pull request merged**